boltdesignsystem/bolt
# [6.x] DS-823 restore radial gradient

## Jira

https://pegadigitalit.atlassian.net/browse/DS-823

## Summary

Restore and build out background radial gradient

## Details

This PR adds back `radial-gradient` as an option for `fill` in the background component.

A summary of the updates:
1. Added full support for the `focal_point` prop (across both horizontal and vertical axes), including full documentation.
2. Made `radial-gradient` fill always full opacity.  It looks like crap when used with any transparent opacity options, and would probably never have been used this way in any designs.
3. Integrated `radial-gradient` with the theming system.
  a. `radial-gradient` changes colors based on the the `t-bolt-*` class of the closest parent.
  b. `--m-bolt-bg-highlight` and `--m-bolt-bg-lowlight` CSS variables have been added to each theme.  These two variables indicate the two colors that should be used in a background gradient within this theme (based on design direction) and could be re-used in the future.
  c. Documentation for theme compatibility of both `gradient` and `radial-gradient` fill options can be seen at /pattern-lab/?p=components-background-theme-compatibility
  d. Documentation for all possible colors can be seen at /pattern-lab/?p=tests-background-themes

Specific questions for reviewers:
1. (easier question) Should we combine the two pages for background theme compatibility?  I had originally intended to keep the main doc page (/pattern-lab/?p=components-background-theme-compatibility) lean, but it honestly just feels more confusing having two pages.  My latest thought is to just keep the sizes as small as reasonable, then group the different fill types in different tabs, following the example of https://boltdesignsystem.com/pattern-lab/?p=layouts-layout-template.
2. (harder question) After building to this point, it feels kind of wrong to have `radial-gradient` as a background `fill` option.  Specifically:
- It is not like the other fill options, which are actually _overlay_ fill options.
- A background gradient could theoretically be applicable anywhere there is a background color (such as cards/teasers-- this is not just theoretical, but something Matt specifically requested).

I have an alternative to propose: Instead of the `fill` prop, we could use one class to add a gradient background to _any_ element, and a second class to modify the gradient focal point.  Something like the following:
```
<div class="t-bolt-navy u-bolt-gradient u-bolt-gradient--top-right">Hello world</div>
```
```
.u-bolt-gradient--top-right {
  --bolt-radial-gradient-focus: 100% 0%;
}
...

.u-bolt-gradient {
  background-repeat: no-repeat;
  background-size: 100% 100%;
  background-image:
    radial-gradient(
      circle at var(--bolt-radial-gradient-focus),
      var(--m-bolt-bg-highlight) 20%,
      var(--m-bolt-bg-lowlight) 80%
    );
}
```

## How to test

- Most simply, create a background with `fill: radial-gradient`.  A basic example can be seen at /pattern-lab/?p=components-background-fill
- For the new `focal_point` options, see /pattern-lab/?p=components-background-focal-point.
- For theme compatibility, see the documentation at /pattern-lab/?p=components-background-theme-compatibility and /pattern-lab/?p=tests-background-themes should make it clear.

## Release notes

- Added full support for the `focal_point` prop
- Added a gradient and gradient focus utility class that works with the Bolt Theming system
- Added `--m-bolt-bg-highlight` and `--m-bolt-bg-lowlight` CSS variables to the Bolt themes

### Visual changes

- Existing Background Comopoent gradients will display as navy instead of black
